commons-core/src/main/scala/com/avsystem/commons/di/Component.scala

```diff
@@ -0,0 +1,240 @@
+    * The component being referred is extracted as a dependency and initialized before the component that refers to
+    * it. This way multiple dependencies can be initialized in parallel.
+    *
+    * @example
+    * {{{
+    *   class FooService
+    *   class BarService
+    *   class Application(foo: FooService, bar: BarService)
+    *
+    *   object MyComponents extends Components {
+    *     def foo: Component[FooService] = singleton(new FooService)
+    *     def bar: Component[BarService] = singleton(new BarService)
+    *
+    *     // before `app` is initialized, `foo` and `bar` can be initialized in parallel
+    *     def app: Component[Application] = singleton(new Application(foo.ref, bar.ref))
+    *   }
+    * }}}
+    */
+  @compileTimeOnly(".ref can only be used inside code passed to component/singleton(...) macro")
+  def ref: T = sys.error("stub")
+
+  def getIfReady: Option[T] =
+    storage.get.option.flatMap(_.value.map(_.get))
+
+  def dependsOn(moreDeps: Component[_]*): Component[T] =
+    new Component(info, deps ++ moreDeps, create, cachedStorage)
+
+  private[di] def cached[T0 >: T](cachedStorage: AtomicReference[Future[T0]], info: ComponentInfo): Component[T0] =
+    new Component(info, deps, create, Opt(cachedStorage))
+
+  /**
+    * Validates this component by checking its dependency graph for cycles.
+    * A [[DependencyCycleException]] is thrown when a cycle is detected.
+    */
+  def validate(): Unit =
+    Component.validateAll(List(this))
+
+  /**
+    * Forces initialization of this component and its dependencies (in parallel, using given `ExecutionContext`).
+    * Returns a `Future` containing the initialized component value.
+    * NOTE: the component is initialized only once and its value is cached.
+    */
+  def init(implicit ec: ExecutionContext): Future[T] =
+    doInit(Nil, starting = true)
+
+  private def doInit(stack: List[Component[_]], starting: Boolean)(implicit ec: ExecutionContext): Future[T] = {
+    val promise = Promise[T]()
+    if (storage.compareAndSet(null, promise.future)) {
+      if (starting) {
+        validate()
+      }
+      val newStack = this :: stack
+      val resultFuture =
+        Future.traverse(dependencies)(_.doInit(newStack, starting = false))
+          .flatMap(resolvedDeps => create(resolvedDeps)(ec))
+          .recoverNow {
+            case NonFatal(cause) =>
+              throw ComponentInitializationException(this, cause)
+          }
+      promise.completeWith(resultFuture)
+    }
+    storage.get()
+  }
+}
+object Component {
+  def async[T](definition: => T): ExecutionContext => Future[T] =
+    _ => Future.eval(definition)
+
+  private case class DfsPtr(component: Component[_], deps: List[Component[_]])
+
+  def validateAll(components: Seq[Component[_]]): Unit = {
+    val visited = new MHashMap[Component[_], Boolean]
+    components.foreach { component =>
```

Comment:
Since you require one entry point, do you really need to `detectCycles` in every component?

Reply:
This method was created specifically for situations where there is _not_ a single entry point (read: CDM)